src/lib.rs

```diff
@@ -0,0 +1,238 @@
+// Copyright (c) 2019 King's College London created by the Software Development Team
+// <http://soft-dev.org/>
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0>, or the MIT license <LICENSE-MIT or
+// http://opensource.org/licenses/MIT>, or the UPL-1.0 license <http://opensource.org/licenses/UPL>
+// at your option. This file may not be copied, modified, or distributed except according to those
+// terms.
+
+#![feature(alloc_layout_extra)]
+
+use std::alloc::{alloc, Layout};
+use std::mem::{size_of, transmute};
+use std::time::Instant;
+
+const REPS: usize = 100;
+const VEC_SIZE: usize = 10000000;
+const VAL_NOREAD: usize = 0xdeadbeef;
+const VAL_WITHREAD: usize = 0xFEEDC0DE;
+
+#[inline(never)]
+fn time<F>(mut f: F)
+where
+    F: FnMut(),
+{
+    let before = Instant::now();
+    for _ in 0..REPS {
+        f();
+    }
+    let d = Instant::now() - before;
+    println!("{:?}", d.as_secs() as f64 + d.subsec_nanos() as f64 * 1e-9);
+}
+
+// The trait whose function we'll dynamically look up.
+trait T {
+    fn m(&self) -> usize;
+}
+
+// This is a simple helper trait that lets us cut a lot of our code in half below: it allows us to
+// instantiate SNoRead or SWithRead via a trait. We can't put this in T because T then wouldn't be
+// a valid trait object.
+trait N {
+    fn new() -> Self;
+}
+
+// SNoRead and SWithRead are the same size, but SNoRead.m() doesn't actually read the "_i"
+// attribute (hence the leading underscore) whereas SWithRead does read from its "i" attribute
+// (hence the lack of a leading underscore).
+struct SNoRead {
+    _i: usize,
+}
+
+impl T for SNoRead {
+    fn m(&self) -> usize {
+        VAL_NOREAD
+    }
+}
+
+impl N for SNoRead {
+    fn new() -> Self {
+        SNoRead { _i: VAL_NOREAD }
+    }
+}
+
+struct SWithRead {
+    i: usize,
+}
+
+impl T for SWithRead {
+    fn m(&self) -> usize {
+        self.i
+    }
+}
+
+impl N for SWithRead {
+    fn new() -> Self {
+        SWithRead { i: VAL_WITHREAD }
+    }
+}
+
+fn vec_normal<S: 'static + N + T>() -> Vec<Box<dyn T>> {
+    let mut v = Vec::<Box<dyn T>>::with_capacity(VEC_SIZE);
+    for _ in 0..VEC_SIZE {
+        v.push(Box::new(S::new()));
+    }
+    v
+}
+
+pub fn bench_normal_no_read() {
+    assert_eq!(size_of::<Box<()>>(), size_of::<usize>());
+    assert_eq!(size_of::<Box<dyn T>>(), size_of::<usize>() * 2);
+    let v = vec_normal::<SNoRead>();
+    time(|| {
+        for e in &v {
+            assert_eq!(e.m(), VAL_NOREAD);
+        }
+    });
+}
+
+pub fn bench_normal_with_read() {
+    assert_eq!(size_of::<Box<()>>(), size_of::<usize>());
+    assert_eq!(size_of::<Box<dyn T>>(), size_of::<usize>() * 2);
+    let v = vec_normal::<SWithRead>();
+    time(|| {
+        for e in &v {
+            assert_eq!(e.m(), VAL_WITHREAD);
+        }
+    });
+}
+
+fn vec_multiref<S: 'static + N + T>() -> Vec<*const dyn T> {
+    vec![Box::into_raw(Box::new(S::new())); VEC_SIZE]
```

Comment:
Would it also be interesting to have tests which "multi-alias" references to the same box?

Reply:
That is very literally what this test is doing ;)

Reply:
Unless I got it wrong, this test operates on raw pointers, not references.

Reply:
Yes, but if we shared a reference multiple times into the `Vec`, we'd really struggle to get it through the borrow checker (e.g. the `Box` would be dropped at the end of this function). I'm not sure it would change anything, as, at run-time, pointers and references are more-or-less the same thing.

[You have made me wonder something else though! Does `vec![x; y]` evaluate `x` then clone the result `y` times, or does it evaluate `x` `y` times?]

Reply:
Aha!

> This will use clone to duplicate an expression, so one should be careful using this with types having a nonstandard Clone implementation. For example, vec![Rc::new(1); 5] will create a vector of five references to the same boxed integer value, not five references pointing to independently boxed integers.

Which is a bit weirdly phrased, but means it's doing the first of the two possibilities.

Reply:
> the `Box` would be dropped at the end of this function

That's a good point. into_raw() hand over freeing responsibilities, so I think the code you have is safe in this regard.